flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/CorrelateSortToRankRule.scala

```diff
@@ -0,0 +1,186 @@
+ * be a Sort with an equal Filter predicate whose keys are same with the LHS grouping keys.
+ *
+ * <p>This rule can only be used in HepPlanner.
+ */
+class CorrelateSortToRankRule extends RelOptRule(
+  operand(classOf[Correlate],
+    operand(classOf[Aggregate],
+      operand(classOf[Project], any())),
+    operand(classOf[Sort],
+      operand(classOf[Project],
+        operand(classOf[Filter], any())))),
+  "CorrelateSortToRankRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val correlate: Correlate = call.rel(0)
+    if (correlate.getJoinType != JoinRelType.INNER) {
+      return false
+    }
+    val agg: Aggregate = call.rel(1)
+    if (agg.getAggCallList.size() > 0
+      || agg.getGroupSets.size() > 1
+      || agg.getGroupSet.cardinality() != 1) {
+      // only one group field is supported now
+      return false
+    }
+    val aggInput: Project = call.rel(2)
+    if (!aggInput.isMapping) {
+      return false
+    }
+    val sort: Sort = call.rel(3)
+    if (sort.offset != null || sort.fetch == null) {
+      // 1. we can not describe the offset using rank
+      // 2. there is no need to transform to rank if no fetch limit
+      return false
+    }
+    val sortInput: Project = call.rel(4)
+    if (!sortInput.isMapping) {
+      return false
+    }
+    val filter: Filter = call.rel(5)
+    val condition = filter.getCondition
+    if (condition.getKind != SqlKind.EQUALS) {
+      return false
+    }
+    // only support one partition key
+    val (inputRef, fieldAccess) = resolveFilterCondition(condition)
+    if (inputRef == null) {
+      return false
+    }
+    val variable = fieldAccess.getReferenceExpr.asInstanceOf[RexCorrelVariable]
+    if (!variable.id.equals(correlate.getCorrelationId)) {
+      return false
+    }
+    aggInput.getInput.getDigest.equals(filter.getInput.getDigest)
+  }
+
+  /**
+   * Resolves the filter condition with specific pattern: input ref and field access.
+   *
+   * @param condition The join condition
+   * @return tuple of operands (RexInputRef, RexFieldAccess),
+   *         or null if the pattern does not match
+   */
+  def resolveFilterCondition(condition: RexNode): (RexInputRef, RexFieldAccess) = {
+    val condCall = condition.asInstanceOf[RexCall]
+    val operand0 = condCall.getOperands.get(0)
+    val operand1 = condCall.getOperands.get(1)
+    if (operand0.isA(SqlKind.INPUT_REF) && operand1.isA(SqlKind.FIELD_ACCESS)) {
+      (operand0.asInstanceOf[RexInputRef], operand1.asInstanceOf[RexFieldAccess])
+    } else if (operand0.isA(SqlKind.FIELD_ACCESS) && operand1.isA(SqlKind.INPUT_REF)) {
+      (operand1.asInstanceOf[RexInputRef], operand0.asInstanceOf[RexFieldAccess])
+    } else {
+      (null, null)
+    }
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val oriBuilder = call.builder()
+    val builder = FlinkRelBuilder.of(oriBuilder.getCluster, oriBuilder.getRelOptSchema)
```

Comment:
nit: one simply way is: using `FlinkRelFactories.FLINK_REL_BUILDER`  to construct `RelOptRule`, and then we can cast `call.builder()` as `FlinkRelBuilder`. one example is `FlinkSubQueryRemoveRule`